File: frontend/src/services/riskEngine.ts
import { RiskLevel, Decision, ToolCallParameters, InjectionResult, RiskFactor } from '../types';

const DANGEROUS_TOOL_PATTERNS = /(?:drop_database|delete_database|delete_table|truncate_table|drop_table|delete_file|remove_file|\bunlink\b|execute_shell|run_command|exec_command|shell_exec|run_script|execute_script|export_credentials|exfiltrate|dump_database|change_permissions|\bchmod\b|\bchown\b|format_disk|wipe_disk|disable_security|bypass_auth|send_email_bulk|mass_delete|bulk_delete)/i;

const CREDENTIAL_PARAM_PATTERNS = /(password|passwd|api[_\-]?key|secret|token|credential|private[_\-]?key|access[_\-]?key|auth[_\-]?token|ssh[_\-]?key|bearer|passphrase|pin|otp)/i;

const EXTERNAL_URL_PATTERNS = /https?:\/\/[^\s"'<>]+/i;

const FILESYSTEM_PATH_PATTERNS = /(\/etc\/|\/var\/|\/home\/|\/root\/|\/sys\/|\/proc\/|\/tmp\/|C:\\Windows\\|C:\\Users\\|C:\\Program|\.\.[\\\/]|\/etc\/passwd|\/etc\/shadow|\.ssh\/|\.aws\/|\.env)/i;

const KNOWN_SAFE_TOOLS = new Set([
  'read_file', 'list_files', 'get_user_info', 'fetch_weather',
  'search_web', 'get_time', 'get_date', 'calculate', 'translate',
  'summarize', 'get_news', 'send_message', 'create_note',
  'get_calendar', 'set_reminder', 'get_stock_price',
]);

const INJECTION_PATTERNS: { category: string; regex: RegExp; points: number }[] = [
  {
    category: 'Instruction Override Attempt',
    regex: /(ignore\s+(all\s+)?(previous|prior|above)\s+instructions|disregard\s+(all\s+)?(previous|prior)\s+rules|forget\s+all\s+previous\s+prompts|new\s+instructions\s+follow|start\s+fresh\s+now)/i,
    points: 35,
  },
  {
    category: 'System Prompt Extraction',
    regex: /(reveal|display|output|print|show|leak)\s+(your\s+)?(system\s+prompt|initial\s+instructions|system\s+message|internal\s+configuration|hidden\s+prompt)/i,
    points: 30,
  },
  {
    category: 'Jailbreak / Unrestricted Persona',
    regex: /(DAN\s+mode|jailbreak|developer\s+mode\s+enabled|act\s+as\s+an\s+unrestricted\s+AI|pretend\s+you\s+have\s+no\s+safety\s+filters|bypass\s+all\s+guidelines)/i,
    points: 40,
  },
  {
    category: 'Security Control Bypass',
    regex: /(bypass\s+(auth|security|filter|guardrail)|disable\s+safety\s+checks|override\s+policy|simulate\s+admin\s+privileges)/i,
    points: 35,
  },
  {
    category: 'Direct Credential Extraction',
    regex: /(give\s+me\s+the\s+(password|api\s*key|root\s*token)|dump\s+all\s+secrets|extract\s+environment\s+variables)/i,
    points: 35,
  },
];

function extractParamInfo(parameters: ToolCallParameters): { keys: string[]; text: string } {
  const keys: string[] = [];
  const values: string[] = [];

  function recurse(obj: any, depth = 0) {
    if (depth > 5 || obj === null || obj === undefined) {
      values.push(String(obj));
      return;
    }
    if (typeof obj === 'object') {
      if (Array.isArray(obj)) {
        for (const item of obj) {
          recurse(item, depth + 1);
        }
      } else {
        for (const [k, v] of Object.entries(obj)) {
          keys.push(k);
          recurse(v, depth + 1);
        }
      }
    } else {
      values.push(String(obj));
    }
  }

  recurse(parameters);
  return { keys, text: values.join(' ') };
}

export function detectPromptInjection(toolName: string, parameters: ToolCallParameters): InjectionResult {
  const { text } = extractParamInfo(parameters);
  const targetText = `${toolName} ${text}`;

  const matchedPatterns: string[] = [];
  let scoreContribution = 0;

  for (const item of INJECTION_PATTERNS) {
    if (item.regex.test(targetText)) {
      matchedPatterns.push(item.category);
      scoreContribution = Math.max(scoreContribution, item.points);
    }
  }

  return {
    detected: matchedPatterns.length > 0,
    patterns_matched: matchedPatterns,
    score_contribution: matchedPatterns.length > 0 ? (scoreContribution || 30) : 0,
  };
}

export function evaluateToolCall(toolName: string, parameters: ToolCallParameters, agentId = 'agent-runtime') {
  const factors: RiskFactor[] = [];
  const reasons: string[] = [];
  const threatSignals: string[] = [];
  const matchedPolicies: string[] = [];

  let rawScore = 0;

  // 1. Dangerous action pattern in tool name
  if (DANGEROUS_TOOL_PATTERNS.test(toolName)) {
    const points = 65;
    rawScore += points;
    factors.push({ points, reason: `Dangerous action keyword detected in tool name: '${toolName}'`, category: 'tool' });
    reasons.push(`[+${points}] Dangerous action keyword: '${toolName}'`);
    threatSignals.push('Destructive Tool Invocation');
    matchedPolicies.push('DESTRUCTIVE_OPERATIONS_POLICY');
  } else if (!KNOWN_SAFE_TOOLS.has(toolName.toLowerCase())) {
    const points = 10;
    rawScore += points;
    factors.push({ points, reason: `Tool '${toolName}' is not in the known-safe catalog`, category: 'tool' });
    reasons.push(`[+${points}] Unknown/Unverified tool: '${toolName}'`);
  }

  // 2. Parameter scanning
  const { keys, text } = extractParamInfo(parameters);

  for (const key of keys) {
    if (CREDENTIAL_PARAM_PATTERNS.test(key)) {
      const points = 25;
      rawScore += points;
      factors.push({ points, reason: `Credential-sensitive parameter detected: '${key}'`, category: 'param' });
      reasons.push(`[+${points}] Credential parameter match: '${key}'`);
      threatSignals.push('Credential Exposure Risk');
      matchedPolicies.push('CREDENTIAL_PROTECTION_POLICY');
      break;
    }
  }

  if (EXTERNAL_URL_PATTERNS.test(text)) {
    const points = 20;
    rawScore += points;
    factors.push({ points, reason: 'External network URL found in parameter payload', category: 'network' });
    reasons.push(`[+${points}] Outbound URL detected in arguments`);
    threatSignals.push('External Network Egress');
    matchedPolicies.push('EXTERNAL_EXFILTRATION_MONITOR');
  }

  if (FILESYSTEM_PATH_PATTERNS.test(text)) {
    const points = 15;
    rawScore += points;
    factors.push({ points, reason: 'Restricted filesystem or sensitive path detected in arguments', category: 'filesystem' });
    reasons.push(`[+${points}] System/Filesystem path parameter match`);
    threatSignals.push('Filesystem Traversal/Access');
    matchedPolicies.push('SYSTEM_RESOURCE_ISOLATION');
  }

  // 3. Prompt injection detection
  const injection = detectPromptInjection(toolName, parameters);
  if (injection.detected) {
    const points = injection.score_contribution;
    rawScore += points;
    for (const pattern of injection.patterns_matched) {
      factors.push({ points, reason: `Prompt injection pattern detected: '${pattern}'`, category: 'injection' });
      reasons.push(`[+${points}] Prompt injection signature: '${pattern}'`);
    }
    threatSignals.push('Prompt Injection Adversarial Pattern');
    matchedPolicies.push('PROMPT_INJECTION_SHIELD');
  }

  const finalScore = Math.min(100, Math.max(0, rawScore));

  let riskLevel: RiskLevel;
  let decision: Decision;

  if (injection.detected) {
    riskLevel = 'RED';
    decision = 'BLOCKED';
    reasons.push('POLICY DECISION: Adversarial Prompt Injection detected → FORCED RED → AUTO-BLOCKED');
  } else if (finalScore <= 30) {
    riskLevel = 'GREEN';
    decision = 'ALLOWED';
    reasons.push('POLICY DECISION: Risk score ≤ 30 → GREEN → AUTO-ALLOWED by AgentGuard');
  } else if (finalScore <= 60) {
    riskLevel = 'AMBER';
    decision = 'PENDING_APPROVAL';
    reasons.push('POLICY DECISION: Risk score 31–60 → AMBER → PENDING HUMAN APPROVAL');
  } else {
    riskLevel = 'RED';
    decision = 'BLOCKED';
    reasons.push('POLICY DECISION: Risk score > 60 → RED → AUTO-BLOCKED by Security Gateway');
  }

  const now = new Date().toISOString();

  return {
    risk_score: finalScore,
    risk_level: riskLevel,
    decision,
    reasons,
    factors,
    threatSignals,
    matchedPolicies,
    injection,
    timestamp: now,
  };
}

File: backend/risk_engine.py
"""
AgentGuard — Risk Engine.

Calculates a deterministic numeric risk score (0–100) based on
structural analysis of the tool call. No LLM involved.

Scoring factors:
  +40  Dangerous action keyword in tool name
  +25  Credential-related parameter names
  +20  External URL in parameter values
  +15  Filesystem / system path in parameter values
  +10  Unknown / unrecognized tool name
  +30  Prompt injection detected (added by policy_engine after injection check)
  [capped at 100]
"""

import re
from typing import Any

# ---------------------------------------------------------------------------
# Dangerous tool name keywords  (exact or substring match, case-insensitive)
# ---------------------------------------------------------------------------
DANGEROUS_ACTION_PATTERNS = re.compile(
    r"(?:"
    r"drop_database|delete_database|delete_table|truncate_table|"
    r"drop_table|delete_file|remove_file|\bunlink\b|"
    r"execute_shell|run_command|exec_command|shell_exec|"
    r"run_script|execute_script|"
    r"export_credentials|exfiltrate|dump_database|"
    r"change_permissions|\bchmod\b|\bchown\b|"
    r"format_disk|wipe_disk|"
    r"disable_security|bypass_auth|"
    r"send_email_bulk|mass_delete|bulk_delete"
    r")",
    re.IGNORECASE,
)

# ---------------------------------------------------------------------------
# Credential-related parameter names
# ---------------------------------------------------------------------------
CREDENTIAL_PARAM_PATTERN = re.compile(
    r"(password|passwd|api[_\-]?key|secret|token|credential|"
    r"private[_\-]?key|access[_\-]?key|auth[_\-]?token|ssh[_\-]?key|"
    r"bearer|passphrase|pin|otp)",
    re.IGNORECASE,
)

# ---------------------------------------------------------------------------
# External URL pattern
# ---------------------------------------------------------------------------
EXTERNAL_URL_PATTERN = re.compile(
    r"https?://[^\s\"'<>]+",
    re.IGNORECASE,
)

# ---------------------------------------------------------------------------
# Filesystem / system path pattern
# ---------------------------------------------------------------------------
FILESYSTEM_PATH_PATTERN = re.compile(
    r"(/etc/|/var/|/home/|/root/|/sys/|/proc/|/tmp/|"
    r"C:\\Windows\\|C:\\Users\\|C:\\Program|"
    r"\.\.[\\/]|"                   # directory traversal
    r"/etc/passwd|/etc/shadow|"
    r"\.ssh/|\.aws/|\.env)",
    re.IGNORECASE,
)

# ---------------------------------------------------------------------------
# Known "safe" tool names that start with 0 base risk
# ---------------------------------------------------------------------------
KNOWN_SAFE_TOOLS = {
    "read_file", "list_files", "get_user_info", "fetch_weather",
    "search_web", "get_time", "get_date", "calculate", "translate",
    "summarize", "get_news", "send_message", "create_note",
    "get_calendar", "set_reminder", "get_stock_price",
}


class RiskFactors:
    """Container for individual risk factor scores and reasons."""

    def __init__(self):
        self.score: int = 0
        self.reasons: list[str] = []

    def add(self, points: int, reason: str) -> None:
        self.score += points
        self.reasons.append(f"[+{points}] {reason}")

    def total(self) -> int:
        return min(self.score, 100)


def score_tool_call(tool_name: str, parameters: dict[str, Any]) -> tuple[int, list[str]]:
    """
    Compute a risk score for a tool call.

    Args:
        tool_name: Name of the tool being called.
        parameters: Parameters dict.

    Returns:
        (risk_score 0–100, list_of_reasons)
    """
    factors = RiskFactors()

    # 1. Dangerous action keyword in tool name
    if DANGEROUS_ACTION_PATTERNS.search(tool_name):
        factors.add(65, f"Dangerous action keyword detected in tool name: '{tool_name}'")

    # 2. Unknown tool
    if tool_name.lower() not in KNOWN_SAFE_TOOLS and not DANGEROUS_ACTION_PATTERNS.search(tool_name):
        factors.add(10, f"Tool '{tool_name}' is not in the known-safe list")

    # Flatten params for text scanning
    param_keys, param_values_text = _extract_param_info(parameters)

    # 3. Credential-related parameter names
    for key in param_keys:
        if CREDENTIAL_PARAM_PATTERN.search(key):
            factors.add(25, f"Credential-sensitive parameter name detected: '{key}'")
            break  # count once per call

    # 4. External URL in parameter values
    if EXTERNAL_URL_PATTERN.search(param_values_text):
        factors.add(20, "External URL found in parameter values")

    # 5. Filesystem / system path in parameter values
    if FILESYSTEM_PATH_PATTERN.search(param_values_text):
        factors.add(15, "System/filesystem path detected in parameter values")

    return factors.total(), factors.reasons


def _extract_param_info(parameters: dict) -> tuple[list[str], str]:
    """
    Recursively extract all parameter keys and stringify all values.

    Returns:
        (list_of_all_keys, concatenated_values_string)
    """
    keys: list[str] = []
    values: list[str] = []
    _recurse(parameters, keys, values, depth=0)
    return keys, " ".join(values)


def _recurse(obj: Any, keys: list, values: list, depth: int) -> None:
    if depth > 5:
        values.append(str(obj))
        return
    if isinstance(obj, dict):
        for k, v in obj.items():
            keys.append(str(k))
            _recurse(v, keys, values, depth + 1)
    elif isinstance(obj, list):
        for item in obj:
            _recurse(item, keys, values, depth + 1)
    else:
        values.append(str(obj))
